// @sap/fesr-to-otel-js/dist/lib/passport/model/ComponentTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ComponentReverseTypes = exports.ComponentTypes = void 0;
var ComponentTypes;
(function (ComponentTypes) {
    ComponentTypes["UNDEFINED"] = "0000";
    ComponentTypes["WEBAS"] = "0001";
    ComponentTypes["J2EE"] = "0002";
    ComponentTypes["TREX"] = "0003";
    ComponentTypes["ICM"] = "0004";
    ComponentTypes["GATEWAY"] = "0005";
    ComponentTypes["CPIC"] = "0006";
    ComponentTypes["BROWSER"] = "0007";
    ComponentTypes["TRACELIB"] = "0008";
    ComponentTypes["DOTNET"] = "0009";
    ComponentTypes["ECLIPSE"] = "000A";
    ComponentTypes["SCPSAP"] = "0014";
    ComponentTypes["PISAP"] = "0015";
    ComponentTypes["SCPNONSAP"] = "0016";
    ComponentTypes["PINONSAP"] = "0017";
    ComponentTypes["SAPPARTNER"] = "0018";
    ComponentTypes["SCPREQUEST"] = "0019";
    ComponentTypes["S4"] = "001E";
    ComponentTypes["SFSF"] = "001F";
    ComponentTypes["ARIBA"] = "0020";
    ComponentTypes["CONCUR"] = "0021";
    ComponentTypes["FIELDGLASS"] = "0022";
    ComponentTypes["CALLIDUS"] = "0023";
    ComponentTypes["BYD"] = "0024";
    ComponentTypes["IBP"] = "0026";
    ComponentTypes["HYBRIS"] = "0027";
    ComponentTypes["SMB"] = "0028";
    ComponentTypes["INDUSTRYCLOUD"] = "0029";
    ComponentTypes["LEONARDO"] = "002A";
    ComponentTypes["CUSTOMERCHECKOUT"] = "002B";
    ComponentTypes["COPILOT"] = "002C";
    ComponentTypes["DWC"] = "0442";
})(ComponentTypes = exports.ComponentTypes || (exports.ComponentTypes = {}));
var ComponentReverseTypes;
(function (ComponentReverseTypes) {
    ComponentReverseTypes["0000"] = "UNDEFINED";
    ComponentReverseTypes["0001"] = "WEBAS";
    ComponentReverseTypes["0002"] = "J2EE";
    ComponentReverseTypes["0003"] = "TREX";
    ComponentReverseTypes["0004"] = "ICM";
    ComponentReverseTypes["0005"] = "GATEWAY";
    ComponentReverseTypes["0006"] = "CPIC";
    ComponentReverseTypes["0007"] = "BROWSER";
    ComponentReverseTypes["0008"] = "TRACELIB";
    ComponentReverseTypes["0009"] = "DOTNET";
    ComponentReverseTypes["000A"] = "ECLIPSE";
    ComponentReverseTypes["0014"] = "SCPSAP";
    ComponentReverseTypes["0015"] = "PISAP";
    ComponentReverseTypes["0016"] = "SCPNONSAP";
    ComponentReverseTypes["0017"] = "PINONSAP";
    ComponentReverseTypes["0018"] = "SAPPARTNER";
    ComponentReverseTypes["0019"] = "SCPREQUEST";
    ComponentReverseTypes["001E"] = "S4";
    ComponentReverseTypes["001F"] = "SFSF";
    ComponentReverseTypes["0020"] = "ARIBA";
    ComponentReverseTypes["0021"] = "CONCUR";
    ComponentReverseTypes["0022"] = "FIELDGLASS";
    ComponentReverseTypes["0023"] = "CALLIDUS";
    ComponentReverseTypes["0024"] = "BYD";
    ComponentReverseTypes["0026"] = "IBP";
    ComponentReverseTypes["0027"] = "HYBRIS";
    ComponentReverseTypes["0028"] = "SMB";
    ComponentReverseTypes["0029"] = "INDUSTRYCLOUD";
    ComponentReverseTypes["002A"] = "LEONARDO";
    ComponentReverseTypes["002B"] = "CUSTOMERCHECKOUT";
    ComponentReverseTypes["002C"] = "COPILOT";
    ComponentReverseTypes["0442"] = "DWC";
})(ComponentReverseTypes = exports.ComponentReverseTypes || (exports.ComponentReverseTypes = {}));
