// @sap/xotel-agent-ext-js/dist/utils/Constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Environment = exports.VisibilityLevel = exports.UseCase = exports.MetricName = exports.AGG_USER = exports.AGG_THRESHOLD = exports.TRACE_STATE_SAPPP = exports.CALM_SERVICE_TYPE = exports.CALM_SERVICE_NAME = exports.FESR_SPAN_NAME = exports.CALM_ATT_TENANCY_TENANT_ID = exports.CALM_ATT_SERVICE_TYPE = exports.CALM_ATT_TENANT_ID = exports.CALM_ATT_ACTION_VERSION = exports.CALM_ATT_ACTION_CONTENT = exports.CALM_ATT_ACTION = exports.CALM_ATT_USE_CASE = exports.ATT_CF_SPACE_NAME = exports.ATT_VISIBILITY_LEVEL = exports.ATT_ACCOUNT = exports.ATT_OUTBOUND_CALLS = exports.ATT_DB_TIME = exports.ATT_DB_CALLS = exports.ATT_EXT_TIME = exports.ATT_EXT_CALLS = exports.ATT_RECEIVED_BYTES = exports.ATT_RESP_TIME_MIN = exports.ATT_RESP_TIME_MAX = exports.ATT_RESP_TIME = exports.ATT_COUNT = exports.ATT_USER_ID = exports.ATT_TRANS_ID = exports.ATT_ROOT_CTX_ID = exports.ATT_PP_ACTION = exports.ATT_HTTP_PASSPORT = exports.ATT_SUBDOMAIN = exports.ORIGIN = exports.CORRELATION_ID_HEADER = exports.PASSPORT_HEADER = exports.PASSPORT_KEY = void 0;
const pjson = require('../../package.json');
exports.PASSPORT_KEY = "passport";
exports.PASSPORT_HEADER = "sap-passport";
exports.CORRELATION_ID_HEADER = "X-CorrelationID";
exports.ORIGIN = "origin";
exports.ATT_SUBDOMAIN = "subdomain";
exports.ATT_HTTP_PASSPORT = "http.request.header.sap_passport";
exports.ATT_PP_ACTION = "pPaction";
exports.ATT_ROOT_CTX_ID = "rootContextId";
exports.ATT_TRANS_ID = "transactionId";
exports.ATT_USER_ID = "userId";
exports.ATT_COUNT = "count";
exports.ATT_RESP_TIME = "responseTime";
exports.ATT_RESP_TIME_MAX = "responseTimeMax";
exports.ATT_RESP_TIME_MIN = "responseTimeMin";
exports.ATT_RECEIVED_BYTES = "receivedBytes";
exports.ATT_EXT_CALLS = "externalCalls";
exports.ATT_EXT_TIME = "externalTime";
exports.ATT_DB_CALLS = "dbCalls";
exports.ATT_DB_TIME = "dbTime";
exports.ATT_OUTBOUND_CALLS = "outboundCalls";
exports.ATT_ACCOUNT = "account";
exports.ATT_VISIBILITY_LEVEL = "sap.visibility.level";
exports.ATT_CF_SPACE_NAME = "sap.cf.space_name";
exports.CALM_ATT_USE_CASE = "sap.calm.useCase";
exports.CALM_ATT_ACTION = "sap.calm.action";
exports.CALM_ATT_ACTION_CONTENT = "sap.calm.actionContent";
exports.CALM_ATT_ACTION_VERSION = "sap.calm.actionVersion";
/**
 * @deprecated use {@link CALM_ATT_TENANCY_TENANT_ID} instead
 */
exports.CALM_ATT_TENANT_ID = "sap.calm.tenantId";
exports.CALM_ATT_SERVICE_TYPE = "sap.calm.serviceType";
exports.CALM_ATT_TENANCY_TENANT_ID = "sap.tenancy.tenant_id";
exports.FESR_SPAN_NAME = "fesr";
exports.CALM_SERVICE_NAME = `${process.env.SAP_CALM_SERVICE_NAME || "otel-app"} - ${pjson.version}`;
exports.CALM_SERVICE_TYPE = process.env.SAP_CALM_SERVICE_TYPE || "SAP_CP_CF";
exports.TRACE_STATE_SAPPP = "sappp";
exports.AGG_THRESHOLD = "SAP_CALM_DCI_AGG_THRESHOLD";
exports.AGG_USER = "SAP_CALM_DCI_AGG_USER";
var MetricName;
(function (MetricName) {
    MetricName["PROCESS_CPU_TIME"] = "process.cpu.time";
    MetricName["PROCESS_CPU_UTILIZATION"] = "process.cpu.utilization";
    MetricName["SYSTEM_FILE_SYSTEM_UTILIZATION"] = "system.filesystem.utilization";
    MetricName["SYSTEM_FILE_SYSTEM_USAGE"] = "system.filesystem.usage";
    MetricName["SYSTEM_MEMORY_USAGE"] = "system.memory.usage";
    MetricName["SYSTEM_MEMORY_UTILIZATION"] = "system.memory.utilization";
})(MetricName = exports.MetricName || (exports.MetricName = {}));
var UseCase;
(function (UseCase) {
    UseCase["RUM"] = "rum";
    UseCase["HEALTHMON"] = "healthmon";
    UseCase["MOCKRECEIVER"] = "mockreceiver";
    UseCase["MOCKSPANRECEIVER"] = "mockspanreceiver";
    UseCase["IMPUSH"] = "impush";
})(UseCase = exports.UseCase || (exports.UseCase = {}));
var VisibilityLevel;
(function (VisibilityLevel) {
    VisibilityLevel["CONFIDENTIAL"] = "confidential";
    VisibilityLevel["INTERNAL"] = "internal";
    VisibilityLevel["PARTNER_RELEVANT"] = "partner_relevant";
    VisibilityLevel["PUBLIC"] = "public";
    VisibilityLevel["CUSTOMER_RELEVANT"] = "customer_relevant";
})(VisibilityLevel = exports.VisibilityLevel || (exports.VisibilityLevel = {}));
var Environment;
(function (Environment) {
    Environment["CF"] = "CF";
    Environment["KYMA"] = "KYMA";
})(Environment = exports.Environment || (exports.Environment = {}));
