// @sap/xotel-agent-ext-js/dist/exporter/ms/MSExporter.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MSExporter = void 0;
const core_1 = require("@opentelemetry/core");
const logger_1 = require("../../common/logger");
const NoopSpanExporter_1 = require("../NoopSpanExporter");
const constants = __importStar(require("../../utils/Constants"));
const MSReadableSpan_1 = require("../../model/ms/MSReadableSpan");
const semantic_conventions_1 = require("@opentelemetry/semantic-conventions");
const resources_1 = require("@opentelemetry/resources");
const TenantKey_1 = require("../../utils/TenantKey");
const INTERVAL = 60 * 1000;
class MSExporter {
    constructor() {
        this.buffer = new Map();
        this._exporter = new NoopSpanExporter_1.NoopSpanExporter();
        this.shutdownOnce = new core_1.BindOnceFuture(this.doShutDown, this);
    }
    set exporter(ex) {
        this._exporter = ex;
    }
    shutdown() {
        return this.shutdownOnce.call();
    }
    clearTimer() {
        if (this.timer !== undefined) {
            clearTimeout(this.timer);
            this.timer = undefined;
        }
    }
    doShutDown() {
        return Promise
            .resolve()
            .then(() => this.flush())
            .then(() => this._exporter.shutdown());
    }
    flush() {
        this.clearTimer();
        const promises = [];
        const newBuffer = [];
        this.buffer.forEach((value, key) => {
            const singles = value;
            const tenantKey = JSON.parse(key);
            let resource = new resources_1.Resource({
                [semantic_conventions_1.SemanticResourceAttributes.SERVICE_NAME]: constants.CALM_SERVICE_NAME,
                [constants.CALM_ATT_SERVICE_TYPE]: constants.CALM_SERVICE_TYPE,
                [constants.CALM_ATT_USE_CASE]: tenantKey.useCase,
                [constants.CALM_ATT_TENANT_ID]: tenantKey.tenantId,
                [constants.CALM_ATT_TENANCY_TENANT_ID]: tenantKey.tenantId,
                [constants.ATT_SUBDOMAIN]: tenantKey.subdomain,
                [constants.CALM_ATT_ACTION_CONTENT]: "",
                [constants.CALM_ATT_ACTION_VERSION]: "",
                [constants.CALM_ATT_ACTION]: "",
            });
            singles.forEach(span => {
                newBuffer.push(new MSReadableSpan_1.MSReadableSpan(span, resource));
            });
        });
        logger_1.logger.debug(`MockExporter: flushing ${newBuffer.length} spans`);
        promises.push(new Promise(this.exportCallbackForMockService(newBuffer)));
        return Promise.all(promises)
            .then(() => {
            this.clearTimer();
            this.buffer.clear();
            Promise.resolve();
        })
            .catch(err => Promise.reject(err));
    }
    groupByTenantForMs(msSpans) {
        const map = new Map();
        msSpans.forEach(eachSpan => {
            const tenantId = makeTenantKey(eachSpan);
            const tenantKey = JSON.stringify(tenantId);
            let spans = map.get(tenantKey);
            if (!spans) {
                spans = [];
            }
            spans.push(eachSpan);
            map.set(tenantKey, spans);
        });
        map.forEach((value, key) => {
            let arr = this.buffer.get(key);
            if (!arr) {
                arr = [];
            }
            arr.push(...value);
            this.buffer.set(key, arr);
        });
    }
    exportCallbackForMockService(msSpans) {
        return (resolve, reject) => {
            this._exporter.export(msSpans, result => {
                var _a;
                if (result.code === core_1.ExportResultCode.SUCCESS) {
                    resolve();
                }
                else {
                    reject((_a = result.error) !== null && _a !== void 0 ? _a : new Error('MockExporter: span export failed'));
                }
            });
        };
    }
    export(spans, resultCallback) {
        if (this.shutdownOnce.isCalled) {
            return;
        }
        logger_1.logger.debug('MockExporter: exporting spans');
        this.groupByTenantForMs(spans);
        this.schedule();
        resultCallback({ code: core_1.ExportResultCode.SUCCESS });
    }
    schedule() {
        if (this.timer !== undefined) {
            return;
        }
        this.timer = setTimeout(() => __awaiter(this, void 0, void 0, function* () {
            try {
                yield this.flush();
            }
            catch (error) {
                console.error(error);
            }
        }), INTERVAL);
        this.timer.unref();
    }
}
exports.MSExporter = MSExporter;
function makeTenantKey(span) {
    // Making key for spans of same type of - tenantId, usecase, subdomain
    let tenantId = span.attributes[constants.CALM_ATT_TENANT_ID];
    if (tenantId == null) {
        tenantId = span.attributes[constants.CALM_ATT_TENANCY_TENANT_ID];
    }
    const subdomain = span.attributes[constants.ATT_SUBDOMAIN];
    const useCase = span.attributes[constants.CALM_ATT_USE_CASE];
    if (tenantId && typeof tenantId === 'string' && useCase && typeof useCase === 'string') {
        if (subdomain && typeof subdomain === "string") {
            return new TenantKey_1.TenantKey(tenantId, useCase, subdomain);
        }
        return new TenantKey_1.TenantKey(tenantId, useCase);
    }
    return new TenantKey_1.TenantKey("", "");
}
