// @sap/xotel-agent-ext-js/dist/config/MessageCountersForConsistencyCheck.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMap = exports.getProducedMessageCount = exports.getSuccessfullyExportedMessageCount = exports.getTotalFailedMessageCount = exports.getTotalExportedMessageCount = exports.checkConsistency = exports.resetCounters = exports.incrementProducedMessageCounter = exports.incrementFailedExportCounter = exports.incrementSuccessfulExportCounter = void 0;
let producedMessageCount = 0;
let successfullyExportedMessageCount = 0;
let failedExportCount = 0;
let producedMessageCountPerUsecase = new Map();
function incrementSuccessfulExportCounter(size) {
    if (checkForReset(successfullyExportedMessageCount)) {
        resetCounters();
    }
    successfullyExportedMessageCount += size;
}
exports.incrementSuccessfulExportCounter = incrementSuccessfulExportCounter;
function incrementFailedExportCounter(size) {
    if (checkForReset(failedExportCount)) {
        resetCounters();
    }
    failedExportCount += size;
}
exports.incrementFailedExportCounter = incrementFailedExportCounter;
function incrementProducedMessageCounter(size, usecaseName) {
    if (checkForReset(producedMessageCount)) {
        resetCounters();
    }
    producedMessageCount += size;
    putInMap(usecaseName, size);
}
exports.incrementProducedMessageCounter = incrementProducedMessageCounter;
function checkForReset(value) {
    if (Number.MAX_VALUE - value < 100) {
        return true;
    }
    return false;
}
function resetCounters() {
    producedMessageCount = 0;
    successfullyExportedMessageCount = 0;
    failedExportCount = 0;
    producedMessageCountPerUsecase.clear();
}
exports.resetCounters = resetCounters;
function checkConsistency() {
    return producedMessageCount === getTotalExportedMessageCount();
}
exports.checkConsistency = checkConsistency;
function getTotalExportedMessageCount() {
    return successfullyExportedMessageCount + failedExportCount;
}
exports.getTotalExportedMessageCount = getTotalExportedMessageCount;
function getTotalFailedMessageCount() {
    return failedExportCount;
}
exports.getTotalFailedMessageCount = getTotalFailedMessageCount;
function getSuccessfullyExportedMessageCount() {
    return successfullyExportedMessageCount;
}
exports.getSuccessfullyExportedMessageCount = getSuccessfullyExportedMessageCount;
function getProducedMessageCount() {
    return producedMessageCount;
}
exports.getProducedMessageCount = getProducedMessageCount;
function getMap() {
    return producedMessageCountPerUsecase;
}
exports.getMap = getMap;
function putInMap(usecaseName, size) {
    if (producedMessageCountPerUsecase.has(usecaseName)) {
        if (producedMessageCountPerUsecase.get(usecaseName) != null) {
            let newValue = producedMessageCountPerUsecase.get(usecaseName) + size;
            producedMessageCountPerUsecase.set(usecaseName, newValue + size);
        }
    }
    else {
        producedMessageCountPerUsecase.set(usecaseName, size);
    }
}
