// @sap/xotel-agent-ext-js/dist/processor/CommonSpanProcessor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommonSpanProcessor = void 0;
const logger_1 = require("../common/logger");
class CommonSpanProcessor {
    constructor(...processors) {
        this.delegates = processors;
    }
    forceFlush() {
        const promises = this.delegates.map((d) => d.forceFlush());
        return Promise
            .all(promises)
            .then(() => {
            logger_1.logger.debug("CommonSpanProcessor: force flushed all span processors");
            return;
        })
            .catch(err => {
            logger_1.logger.error(`CommonSpanProcessor: force flush failed '${err}'`);
            return;
        });
    }
    onStart(span, parentContext) {
        this.delegates.forEach(d => d.onStart(span, parentContext));
    }
    onEnd(span) {
        this.delegates.forEach(d => d.onEnd(span));
    }
    shutdown() {
        const promises = this.delegates.map((d) => d.shutdown());
        return Promise
            .all(promises)
            .then(() => {
            logger_1.logger.debug("CommonSpanProcessor: shut down all span processors");
            return;
        })
            .catch(err => {
            logger_1.logger.error(`CommonSpanProcessor: shut down failed '${err}'`);
            return;
        });
    }
}
exports.CommonSpanProcessor = CommonSpanProcessor;
