// @sap/xotel-agent-ext-js/dist/lib/passport/model/ActionType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActionType = void 0;
var ActionType;
(function (ActionType) {
    ActionType[ActionType["NO_ACTION"] = 0] = "NO_ACTION";
    ActionType[ActionType["TRANSACTION"] = 1] = "TRANSACTION";
    ActionType[ActionType["BATCH"] = 2] = "BATCH";
    ActionType[ActionType["REPORT"] = 3] = "REPORT";
    ActionType[ActionType["WEB_SERVICE"] = 4] = "WEB_SERVICE";
    ActionType[ActionType["UNDEFINED_"] = 15] = "UNDEFINED_";
    ActionType[ActionType["HTTP"] = 11] = "HTTP";
    ActionType[ActionType["HTTP_OPT"] = 12] = "HTTP_OPT";
    ActionType[ActionType["HTTP_HEAD"] = 13] = "HTTP_HEAD";
    ActionType[ActionType["HTTP_VALUE"] = 14] = "HTTP_VALUE";
    ActionType[ActionType["HTTP_PREFETCH"] = 15] = "HTTP_PREFETCH";
    ActionType[ActionType["HTTP_SCROLLING"] = 16] = "HTTP_SCROLLING";
    ActionType[ActionType["HTTP_SRV_PUSH"] = 17] = "HTTP_SRV_PUSH";
})(ActionType = exports.ActionType || (exports.ActionType = {}));
